### PR TITLE
add inventory model

### LINKED ISSUE #11: Create Inventory Model
This will be the model to encapsulate all the items into 1 group. This is the level where user permissions will be applied.
Need a name and description for now, we'll connect it to users later.

### PR DESCRIPTION
**Pull request by Exellin**
resolves #11

### CODE REVIEW

File: app/models/inventory.rb [hunk 1]
```diff
@@ -0,0 +1,4 @@
+class Inventory < ApplicationRecord
+  validates :name, presence: true, length: { maximum: 255 }
+  validates :description, presence: true, length: { maximum: 255 }
+end
```
Comment (pachonk) [inline app/models/inventory.rb]: New line at EOF please.

Comment (Exellin) [inline app/models/inventory.rb]: done for all files

File: spec/models/inventory_spec.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
+require 'rails_helper'
+
+RSpec.describe Inventory, type: :model do
+  it "has a valid factory" do
+    expect(FactoryGirl.build(:inventory)).to be_valid
+  end
+
+  describe "presence" do
+    it "is invalid with an empty name" do
+      inventory = Inventory.create(description: "this is a description")
```
Comment (pachonk) [inline spec/models/inventory_spec.rb]: This test becomes less and less manageable as time moves on. I suggest doing something like this:
`expect(FactoryGirl.build(:inventory, name: nil)).not_to be_valid`

Comment (Exellin) [inline spec/models/inventory_spec.rb]: updated all invalid tests to use factorygirl

File: spec/models/inventory_spec.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
+require 'rails_helper'
+
+RSpec.describe Inventory, type: :model do
+  it "has a valid factory" do
```
Comment (pachonk) [inline spec/models/inventory_spec.rb]: Fair warning: this is me being nit picky. We will need to decide as a team if we want to follow this rule, but I follow it strictly personally.
Technically you should always use single quotes (`'`) unless you need string interpolation. They are faster to type (no need to press shift) as well as (marginally) faster for ruby to execute.
See: https://github.com/bbatsov/ruby-style-guide#strings

File: db/schema.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+# system, you should be using db:schema:load, not running all the migrations
+# from scratch. The latter is a flawed and unsustainable approach (the more migrations
+# you'll amass, the slower it'll run and the greater likelihood for issues).
+#
+# It's strongly recommended that you check this file into your version control system.
+
+ActiveRecord::Schema.define(version: 20170702185054) do
+
+  # These are extensions that must be enabled in order to support this database
+  enable_extension "plpgsql"
```
Comment (pachonk) [inline db/schema.rb]: (note: this is rails gen'd, so these quotes should never be manually changed)

File: spec/models/inventory_spec.rb (hunk 3 above)
Comment (Exellin) [inline spec/models/inventory_spec.rb]: Yeah I agree we should decide on a style for everything in the repo and follow it

Comment (Exellin) [inline spec/models/inventory_spec.rb]: Fixed.

**Comment by pachonk** [comment]
I have requested a review from @brandondees if he has time.
On any note, I'd like 1 more peer review before we pull this in.
Nice job man!

File: db/schema.rb [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+  create_table "inventories", force: :cascade do |t|
+    t.string "name", limit: 255
+    t.string "description", limit: 255
```
Comment (brandondees) [inline db/schema.rb]: if we're validating that these are absolutely required, that should be backed by a not-null constraint in the DB as well. in the real world it's quite common for some DB administrator to want to go manipulate data directly rather than through the application UI which can lead to apps breaking in annoying ways or even hard-to-troubleshoot ways. 

general rule of thumb: database constraints are consistency guarantees, rails validations are merely user guidance.

Comment (brandondees) [inline db/schema.rb]: (can be implemented as a separate issue / migration so as not to block acceptance of this PR, btw)

Comment (pachonk) [inline db/schema.rb]: Hey, good catch. I didn't catch that! I'm SUPER happy you reviewed for this very reason!

File: app/models/inventory.rb [hunk 6]
```diff
@@ -0,0 +1,4 @@
+class Inventory < ApplicationRecord
+  validates :name, presence: true, length: { maximum: 255 }
+  validates :description, presence: true, length: { maximum: 255 }
```
Comment (brandondees) [inline app/models/inventory.rb]: where are the user story requirements for these? i find in real world projects, description in addition to a name / primary identifier is very often considered unnecessary by users and they find it frustrating to require it. will the UI have an explanation for why these are essential, and how to fill them with useful content?

Comment (brandondees) [inline app/models/inventory.rb]: @pachonk @Exellin ☝️

Comment (brandondees) [inline app/models/inventory.rb]: should name also be unique for any reason? perhaps at least scoped within a given tenant / inventory?

Comment (brandondees) [inline app/models/inventory.rb]: are these fields always meaningful to the user, and must they always be hand-chosen at creation time? does it make sense to have a safe default value generated so that users who don't care about these specifics can move along with their workflow more quickly? (I.e. randomly generate a unique name identifier at object instantiation to use as a #new form placeholder)

File: db/migrate/20170702185054_create_inventories.rb [hunk 7]
```diff
@@ -0,0 +1,8 @@
+class CreateInventories < ActiveRecord::Migration[5.1]
+  def change
+    create_table :inventories do |t|
+      t.string :name, :limit => 255
+      t.string :description, :limit => 255
```
Comment (brandondees) [inline db/migrate/20170702185054_create_inventories.rb]: no timestamps? or is that created implicitly now?

File: db/schema.rb (hunk 4 above)
Comment (brandondees) [inline db/schema.rb]: there should probably be an issue to add a default .rubocop.yml and enforce linting for all contributions. rails-generated file locations can be whitelisted for exclusion from the rules

File: spec/models/inventory_spec.rb [hunk 8]
```diff
@@ -0,0 +1,27 @@
+require 'rails_helper'
+
+RSpec.describe Inventory, type: :model do
+  it 'has a valid factory' do
+    expect(FactoryGirl.build(:inventory)).to be_valid
+  end
+
+  describe 'presence' do
```
Comment (brandondees) [inline spec/models/inventory_spec.rb]: is presence a method on the Inventory object? i think this could be structured / worded more clearly

File: spec/models/inventory_spec.rb [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+    it 'is invalid with an empty name' do
+      expect(FactoryGirl.build(:inventory, name: nil)).not_to be_valid
+    end
+
+    it 'is invalid with an empty description' do
```
Comment (brandondees) [inline spec/models/inventory_spec.rb]: it usually makes more sense to group assertions about a given attribute together under a `describe '#attr_name'` block rather than grouping them by common behavior, unless it really makes a big difference in terms of DRY code and is considered worth it.

Comment (brandondees) [inline spec/models/inventory_spec.rb]: see http://www.betterspecs.org/ for a collection of generally good advice for rspec style. i disagree with some points but on average the rationale is clear enough and these rules make a decent baseline to follow by default

File: db/migrate/20170702185054_create_inventories.rb (hunk 7 above)
Comment (pachonk) [inline db/migrate/20170702185054_create_inventories.rb]: Ever since I've been using rails, timestamps are default.

File: app/models/inventory.rb (hunk 6 above)
Comment (pachonk) [inline app/models/inventory.rb]: I agree, let's not require presence of description, but let's require presence and uniq (later scoped to user/org) on name

File: spec/models/inventory_spec.rb (hunk 8 above)
Comment (pachonk) [inline spec/models/inventory_spec.rb]: Maybe we can group the 2 describe blocks and call em "validations"

Comment (brandondees) [inline spec/models/inventory_spec.rb]: that's not my go-to pattern anymore but it works well enough.

File: db/migrate/20170702185054_create_inventories.rb (hunk 7 above)
Comment (Exellin) [inline db/migrate/20170702185054_create_inventories.rb]: added timestamps they weren't created by default with rails g migration.

File: spec/models/inventory_spec.rb (hunk 9 above)
Comment (Exellin) [inline spec/models/inventory_spec.rb]: I updated the specs to be grouped by attribute.  I'll also take a look through that link thanks for the resource.

File: app/models/inventory.rb (hunk 6 above)
Comment (Exellin) [inline app/models/inventory.rb]: updated name to be case sensitive and removed presence validation on description.

File: db/schema.rb (hunk 5 above)
Comment (Exellin) [inline db/schema.rb]: So people go and update data on the database itself instead of through the rails console?  I was under the impression that ui validations were things handled on the browser level which won't let you submit a form until they are all met.  The validations in the model are still always present until you bypass both the ui and the rails console.

If this isn't the case, should I also research a way to validate that names are unique into the schema?

I added the not-null constrain to name and those constraints are automatically applied to timestamps in the meantime.

File: spec/factories/inventory.rb [hunk 10]
```diff
@@ -0,0 +1,8 @@
+require 'faker'
+
+FactoryGirl.define do
+  factory :inventory do
+    name { Faker::Lorem.word }
```
Comment (pachonk) [inline spec/factories/inventory.rb]: I love to use a "fun" faker method that has a lot of uniqueness. Also, use .uniq since we're validating uniqueness.

Comment (pachonk) [inline spec/factories/inventory.rb]: https://github.com/stympy/faker#ensuring-unique-values

Comment (brandondees) [inline spec/factories/inventory.rb]: uniq on faker means you'll run out of values if you test with too many instances. IMHO it's far safer to just concatenate the string with some true random data or use a sequence to include a non-repeating number as part of it

Comment (sanchez2210) [inline spec/factories/inventory.rb]: Does this mean that the file should get modified now? (Adding random string)

Comment (pachonk) [inline spec/factories/inventory.rb]: @sanchez2210 Yes, but let @Exellin do it since it's his PR :dancer:

Comment (pachonk) [inline spec/factories/inventory.rb]: Or use a faker attribute that has MANY values.

File: spec/models/inventory_spec.rb [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+    end
+
+    it 'is invalid with a duplicate value' do 
+      inventory = FactoryGirl.create(:inventory, name: 'duplicate')
+      expect(FactoryGirl.build(:inventory, name: 'Duplicate')).not_to be_valid
+    end
+  end
+
+  describe '#description' do
+    it 'is invalid when over 255 characters' do
```
Comment (pachonk) [inline spec/models/inventory_spec.rb]: You should also explicitly test that it IS valid when it's nil and when it's < 255 char.

Comment (pachonk) [inline spec/models/inventory_spec.rb]: Same for #name

Comment (Exellin) [inline spec/models/inventory_spec.rb]: added those tests

File: spec/models/inventory_spec.rb [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+  describe '#name' do
+    it 'is invalid when nil' do
+      expect(FactoryGirl.build(:inventory, name: nil)).not_to be_valid
+    end
+
+    it 'is invalid when over 255 characters' do
+      expect(FactoryGirl.build(:inventory, name: Faker::Lorem.characters(256))).not_to be_valid
+    end
+
+    it 'is valid when under 256 characters' do
```
Comment (pachonk) [inline spec/models/inventory_spec.rb]: I typically like to group the valid and invalid together. You have a valid test right smack dab in the middle of these tests (same with desc, but with invalid)

Comment (Exellin) [inline spec/models/inventory_spec.rb]: oh I was grouping it so the 256/255 value were right next to each other.

Comment (Exellin) [inline spec/models/inventory_spec.rb]: done

File: spec/factories/inventory.rb [hunk 13]
```diff
@@ -0,0 +1,8 @@
+require 'faker'
```
Comment (pachonk) [inline spec/factories/inventory.rb]: This shouldn't need to exist since it's required from gemfile.
